### ISSUE TITLE
--cache-from prevents sharing layers across containers

### ISSUE THREAD

**Issue by daveisfera** [moby/moby#33002]
**Description**
Before 1.10, you could do a `docker pull` and then a build with the cache being used so that only changed layers were rebuilt. This is no longer possible since the parent relationships are not maintained by the `pull`. Using `--cache-from` allows the pulled image to be used by the cache but this prevents sharing layers across containers.

**Steps to reproduce the issue:**
1. Pull `image_a`
2. Build `image_a` with `--cache-from=image_a` 
3. Build `image_b` with `--cache-from=image_b`

**Describe the results you received:**
`image_b` builds an entirely new image instead of using the layers that should be shared from `image_a`. 

**Describe the results you expected:**
Layers that can be shared between `image_a` and `image_b` are shared.

**Output of `docker version`:**
```
Docker version 17.03.1-ce, build c6d412e
```

**Output of `docker info`:**
```
Containers: 20
 Running: 20
 Paused: 0
 Stopped: 0
Images: 85
Server Version: 17.03.1-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 4ab9917febca54791c5f071a9d1f404867857fcc
runc version: 54296cf40ad8143b62dbcaa1d90e520a2136ddfe
init version: N/A (expected: 949e6facb77383876aeff8a6944dde66b3089574)
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.13-moby
Operating System: Alpine Linux v3.5
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 4.833 GiB
Name: moby
ID: T4IE:7QYK:32DA:MS6V:K6BF:7WBH:GW6Y:VNJK:U65D:XPK5:6KTA:JWD3
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 185
 Goroutines: 156
 System Time: 2017-05-03T20:48:00.005833217Z
 EventsListeners: 1
No Proxy: *.local, 169.254/16
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by tonistiigi** [comment]
@daveisfera Sorry, I don't understand what problem you are describing. What's the difference between imagea and imageb? If you want to use multiple images as a cache source you can specify `--cache-from` multiple times.

**Comment by daveisfera** [comment]
Sorry, here's a few more details.

imagea and imageb are a Python/Node/whatever micro-service so the user's code is the last or one of the last layers to be added. The first few layers are "setup layers" that install system libraries or configure the environment and the bulk of those should reasonably be shared by the images being built. Specifying `--cache-from` multiple times works when it's only 2 images but when you start building 10 or more, it becomes a pain to manage and potentially problematic.

So what happens when I build a new image and forget to specify `--cache-from` for all the previously built ones? I now have duplicated layers that have the exact same content and should be shared but aren't.

And these are just the simple cases, so what happens when you're building a lot of images and knowing which should be shared becomes too difficult to track with a sane amount of effort?

**Comment by tonistiigi** [comment]
@daveisfera So there is no actual issue? Problem is that specifying `--cache-from` is inconvenient when there are very many images. #26065 has some background info. In theory I think we could support `--cache-from=*` if we used time-based priority(same we use for local cache) for matching instead of user specified order( cc @stevvooe).

**Comment by daveisfera** [comment]
@tonistiigi From my perspective it's an issue because it means that there's currently no way to do builds with shared layers in a CI style environment where the cache from previous builds is not available and is being reconstructed using `pull`/`push` or `load`/`save`.

Also, if you skip the `pull` out of the above steps (i.e. just building the two images on a machine), you end up with images that don't share layers when I think most people would say that these two sets of commands should give you identical results (i.e. layers shared in both cases).

For example, these two images will share layers:
```
docker build -t image_a:dev image_a
docker build -t image_b:dev image_b
```

While these two won't:
```
docker build -t image_a:dev --cache-from image_a:dev image_a
docker build -t image_b:dev --cache-from image_a:dev image_b
```

Basically, `--cache-from` is exclusive and the documentation doesn't point that out. The only info it provides is `Images to consider as cache sources` when something like `Consider only these images as cache sources` or `Images to exclusively consider as cache sources` would help users not be caught off guard. Some way to relax that or use `--cache-from=*` would definitely be a big step in the right direction.

**Comment by daveisfera** [comment]
@tonistiigi Also, based on [this comment](https://github.com/moby/moby/issues/26065#issuecomment-249046559) if the order of the multiple `--cache-from` statements wrong then there will be fewer shared layers than expected. So a simple approach of just listing them all won't work because order matters and we're now back to doing efficient builds is very difficult to manage.

**Comment by tonistiigi** [comment]
@daveisfera That's what I meant by the need to switch to time based ordering in https://github.com/moby/moby/issues/33002#issuecomment-299041162 . There are use cases with both methods though, and you can construct an example where user defined order is more correct.

**Comment by daveisfera** [comment]
@tonistiigi That makes sense, but my point was to show another case where the current behavior is unexpected/undesirable to hopefully point out that the original question about there not being an "actual issue" is understood to be a firm no and this is definitely an issue that needs to be resolved as soon as possible.

**Comment by daveisfera** [comment]
Is there an estimate on when `--cache-from=*` or similar functionality to fix this issue will be available?

Also, is there any sort of workaround that could be used currently to allow cache re-use and sharing across images in CI environments? (maybe something like [buildcache](https://github.com/tonistiigi/buildcache) but with support for multiple images)

**Comment by Schnitzel** [comment]
I've got a similar issue, but mine is for Multi Stage Builds and not two Docker Images that share some commands: https://github.com/moby/moby/issues/34715

would be awesome to have `--cache-from` also looking up the local cache

**Comment by daveisfera** [comment]
@tonistiigi Any update on when we can expect to a fix for this? Or at least a workaround?

**Comment by tonistiigi** [comment]
@daveisfera Adding multiple images support to `buildcache` tool should be quite simple but you still have to list them on `save` (like you need to do for `--cache-from`). The real solution should be integrating with https://github.com/moby/buildkit that uses a caching model that could import build cache from remote sources on demand.

**Comment by daveisfera** [comment]
There's a big warning on `buildkit` about it being "not suitable for practical workloads". Any options that are possible now or in the short term?

Also, what about support for `--cache-from=*`? That seems like it would be the simplest and most robust solution.

**Comment by daveisfera** [comment]
We're now past the 1 year anniversary of this been broken, so is there any update on when a solution for this can be expected?

**Comment by docwhat** [comment]
@itayadler How does buildkit help here?  I read [Build Enhancements for Docker for 18.09](https://docs.docker.com/develop/develop-images/build_enhancements) and it didn't mention caching.

**Comment by dimaqq** [comment]
> In docker 18.09 try to remove --cache-from and enable buildkit:
> `export DOCKER_BUILDKIT=1`

@itayadler please explain how to use it 🙏 

In my local testing (18.09.02, Mac) it works worse than without buildkit:
* always rebuilds if `--cache-from` is omitted (same as before)
* errors out if some of `--cache-from X` are not in images (worse UX, can't dump tons of tags hoping for the best)
* appears to pull locally missing tags from docker hub (can be useful, but...)
* fails on buildkit-built `--cache-from` references (see below)

`error on cache query: invalid build cache from {MediaType:application/vnd.docker.distribution.manifest.v2+json Digest:sha256:bd[hash-omitted] Size:1996 URLs:[] Annotations:map[] Platform:<nil>}`
